[![DOI](https://zenodo.org/badge/199330561.svg)](https://zenodo.org/badge/latestdoi/199330561)

# Richman et al. *Cell Systems* 2019
Analysis to generate figures in the AG paper.

# See also
[antigen.garnish](https://github.com/immune-health/antigen.garnish)

# Dependencies from CRAN not installed by *antigen.garnish*
 * [pROC](https://cran.r-project.org/web/packages/pROC/index.html)
 * [gglogo](https://cran.r-project.org/web/packages/gglogo/index.html)
 * [gridExtra](https://cran.r-project.org/web/packages/gridExtra/index.html)
 * [VennDiagram](https://cran.r-project.org/web/packages/VennDiagram/index.html)
 * [microbenchmark](https://cran.r-project.org/web/packages/microbenchmark/index.html)
 * [survival](https://cran.r-project.org/web/packages/survival/index.html)
 * [pheatmap](https://cran.r-project.org/web/packages/pheatmap/index.html)

# Welcome
Each folder contains scripts to conduct the main analysis or generate figures. It is assumed that the working directory is the one containing each respective script. **Scripts are heavily annotated and meant to be run line by line**, not sourced, although they will function that way. Please read annotated code carefully to see notes on parallelization and customizing output. There is sufficient data and code to recreate all of the figures in the manuscript and some additional interesting analyses.

Please contact [@leeprichman](https://github.com/leeprichman) or leave an issue in the repo if you have any trouble, questions, or suggestions!

Thank you for taking the time to replicate our analysis!

# Manifest

 * **Main analysis/**: Scripts to prepare the 5 clinical trial datasets in the paper (Riaz et al., Rizvi et al., Van Allen et al., Snyder et al., Hellmann et al.) and run antigen.garnish to generate the combined output file.  These scripts are written minimally, in reality, it will require careful chunking and parallelization to run the actual analysis in a reasonable timeframe without hitting memory limits. If you wish to repeat the original analysis from the raw data, be aware of segmentation faulting from netMHC and BLAST.  Of note, Biostring's Smith-Waterman alignment will exceed memory and simply drop threads without ever returning an error and will return an incorrect value  dissimilarity or IEDB score for some peptides. If IEDB score or dissimilarity do not appear to match up to data deposited in the repository, this is most likely why. Try reducing parallelization or chunking into smaller jobs in this case.

   + **scripts**:
   + **Get_Hellmann_files.R**: Retrieve VCFs from the European Variation Archive and pass them to antigen.garnish for the Hellmann et al. 2017 study.
   + **Riaz_input.R**: script to parse and process Riaz et al. 2017 dataset from supplementary tables.
   + **VanAllen_Snyder_Rizvi_analysis.R**: script to parse and process the Van Allen et al. 2015, Snyder et al. 2014, and Rizvi et al. 2015 datasets
   + **garnish_venn.R**: custom function to classify neoantigens and calculate overlap meant only for this repository, sourced and called by various scripts in this folder. `antigen.garnish::garnish_summary` performs identical classification and other computations without returning overlap.

   + **files**:
   + *combined_output.txt*: table of neoantigen classifications, overlap, TMB, and OS/PFS for available patients from all five datasets, from combining all other output files
   + *PRJEB24995.txt* and *Hellmann_meta.txt*: File manifest and metadata for the Hellmann dataset from which `garnish_affinity` input will be derived.
   + *Riaz_Alleles.xlsx* and *Riaz_Mutations.xlsx*: supplementary tables from Riaz et al. from which `garnish_affinity` input will be derived
   + *VASR_input.tpxz*: [pixz](https://github.com/vasi/pixz) and tar compressed input for the Van Allen et al., Snyder et al., and Rizvi et al. data, compiled from the original publications and annotated by [SnpEff](http://snpeff.sourceforge.net). Adapted from [Rech et al.  2018](http://cancerimmunolres.aacrjournals.org/content/6/3/276.long). This will be need to be decompressed prior to analysis, see [walkthrough step 2](#walkthrough).
   + *TMB_table.txt*: Complete table of TMB and survival compiled from all five datasets. Hellmann et al. and Rizvi et al. - PFS, Van Allen et al., Riaz et al., and Snyder et al. - OS.
   + *Hellmann_input.txt.xz* and *Riaz_input.txt.xz* and *VASR_input.tpxz*: tarred and/or [xz](https://tukaani.org/xz/) compressed tables of input variants generated by the above scripts, ready for `garnish_affinity`.

 * **Figure 1/**: Ensemble prediction figures.

    + **Fig1BC.R** and **Fig1D.R**: Script to generate figures.
    + *bdata.20130222.mhci.txt*: Total data from [Kim et al. 2014](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-15-241), retrieved from the [IEDB](http://tools.iedb.org/static/main/binding_data_2013.zip).
    + *ensemble_pred_mtable.tpxz*: Compressed output from all MHC affinity predictions on the Kim et al. dataset. This will need to be decompressed if not re-performing the predictions yourself, see [walkthrough step 6](#walkthrough).

 * **Figure 2/**: Partition function validation and immunogenicity analysis.

    + *pnas_chowell_2015.txt*: Input from [Chowell et al. 2015](http://www.pnas.org/content/112/14/E1754.long) taken from [Table S1](http://www.pnas.org/highwire/filestream/618981/field_highwire_adjunct_files/1/pnas.1500973112.sd01.xls). Database of 9,888 experimentally confirmed MHC binders, with ~50% having confirmed immunogenicity by T cell assay.
    + *Chowell_st_ie.txt*: Dissimilarity and IEDB score values for the Chowell et al. dataset.
    + *Chowell_preds.txt*: MHC affinity predictions for all 4 digit HLA resolution entries in the Chowell et al. dataset.
    + *Luksza_input.txt*: Complete input data taken from Luksza et al. 2017 ([Supplementary Data File 7](https://media.nature.com/original/nature-assets/nature/journal/v551/n7681/extref/nature24473-s9.zip))
    + *Luksza_output.txt*: Complete output data taken from Luksza et al. 2017 ([Supplementary Data File 7](https://media.nature.com/original/nature-assets/nature/journal/v551/n7681/extref/nature24473-s9.zip))
    + *parameters_comp.txt*: intermediate output from Fig2A.R script.
    + *\*ROCs.RDS*: intermediate RDS format output from AUC analysis in Fig2CE_S2A.R script.
    + **Fig2A.R**: script to generate Figure 2A, comparison of "iedb_score" and "TCR recognition probability" from [Luksza et al. 2017](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6137806/).
    + **Fig2CE_S2A**: script to generate Figures 2C,E and Figure S2A: receiver-operator characteristic curves for immunogenicity in the Chowell et al. dataset.
    + **Fig2D.R**: script to generate Figure 2D, contingency tables for immunogenicity analysis using Chowell et al. dataset.

 * **Figure 3/**: Characterization of neoantigens identified by multiple quality metrics.

    + **Fig3A_S2B.R**: Distribution of dissimilarity (Fig3A) and IEDB scores (FigS2B).
    + **Fig3CDE.R**, and **Fig3F.R**: Scripts to perform analysis and generate figures. (C) Self-alignment analysis; (D) Sequence Logo analysis; (E) Hydropathy analysis; (F) Venn diagram of neoantigens.
    + *Kyte_Doolittle.txt*: Table of KD hydropathy scores.
    + *all_MHC_binders.fa*: Intermediate fasta file created for self alignment analysis, used to generate Figure 3C.
    + *all_align_lengths.txt*: Output from BLAST alignment figure  analysis (Figure 2C).
    + *all_MHC_binders_w_id.txt*: All neoantigens with MHC affinity < 500nM from the 5 datasets with quality analyses metrics.
    + *all_non_binders_w_names.txt*: All neoantigens with MHC affinity 1000-5000nM from the 5 datasets.
    + Also includes various logo plots for different common 2 digit allele families.

 * **Figure 4/**: TMB correlation and Cox Proportional Hazard model survival analysis.

    + **Fig4A.R**: Script to perform TMB correlation to neoantigen burden in combined data.
    + **All_dissimilarity_9mers.R**: Script to compute dissimilarity from all neopeptides in all 318 patients from the five datasets. Output generates the *Dissimilarity_9mers_patient_level.txt* file.
    + **Fig4BC.R**: Script to perform Cox Proportional Hazard model and generate heatmaps for Figure 4B-C.
    + *Supplementary_Data_1.txt*: Table of survival analysis output, HR, confidence intervals, and p-values for all survival comparisons made.

 * **Figure S/**: Supplemental and additional figures not included in the paper with analyses of interest.

    + **Alt_cutoff_contingency_tables.R**: script to generate contingency tables as in Figure 2C and Figure S2C, with alternative cutoffs for dissimilarity and IEDB score.
    + **FigS1.R**: Script to generate software benchmarking figure.
    + **FigS3D.R**: Script to generate figure of distribution of neoantigens across clinical datasets.
    + **Balachandran_muc16.R**: Dissimilarity analysis for single neoantigen from [Balachandran et al., *Nature* 2017](https://www.ncbi.nlm.nih.gov/pubmed/29132146)
    + **Chheda_H3.3K27M.R**: Dissimilarity analysis for single neoantigen from [Chheda et al., *J Ex Med* 2017](https://www.ncbi.nlm.nih.gov/pubmed/29203539)
    + **Keskin_analysis.R**: Script to analyze vaccines from GBM trial, [Keskin et al., *Nature* 2019](https://www.ncbi.nlm.nih.gov/pubmed/30568305)
    + *41586_2018_792_MOESM5_ESM.xlsx* & *Keskin_vaccines.txt*: Supplemental table and parsed version from Keskin et al.
    + **Le_analysis.R**: analysis of neoantigens from MMR-deficient tumors [Le et al., *Science* 2017](https://www.ncbi.nlm.nih.gov/pubmed/28596308)
    + **Fritsch_Ott.R**: analysis of neoantigens from [Fritsch et al., *Cancer Immunol Res* 2014](https://www.ncbi.nlm.nih.gov/pubmed/24894089) and [Ott et al., *Nature* 2017 ](https://www.ncbi.nlm.nih.gov/pubmed/28678778)
    + *Wu_mel.csv* & *Fritsch_CIR_2014.txt*: Table of neoantigens for the above analysis.
    + additional figures generated as output from figure scripts not included in manuscript such as boxplot of binding affinity of neoantigens by quality metric and proportions of high dissimilarity neos that fall into other groups

# Walkthrough

Intermediate output is provided in most of the folders, so repeating the analysis can be done  at several points.  To recreate everything from scratch:

  1. Download and install antigen.garnish from the shell.  Run test suite to verify install.
```
curl -fsSL http://get.rech.io/antigen.garnish.sh | sudo sh
sudo Rscript -e 'testthat::test_package("antigen.garnish")'

```
  To bootstrap a fresh AWS instance, check out the antigen.garnish [wiki](https://github.com/immune-health/antigen.garnish/wiki/Installation).

  2. Set working directory to Main_analysis/ and run through the following scripts, ensuring appropriate parallelization and memory usage.  See annotated code within scripts for suggestions. This is no small task and will require parallel computing. The output from this step is provided in the repository so feel free to **skip to step 5 and beyond if you are not interested in replicating this part.**

  * Get_Hellmann_files.R

  * Riaz_input.R

  * VanAllen_Snyder_Rizvi_analysis.R - prior to this you will need to decompress the input data, you will first need to install [pixz](https://github.com/vasi/pixz), then:
```
  pixz -d VASR_input.tpxz
  tar -xvf VASR_input.tar

```
  3. Combine `garnish_venn` output from all 3 scripts into a single file. From within R:
```
library(data.table)
library(magrittr)
lapply(c("VASR_output.txt",
          "Riaz_output.txt",
          "Hellmann_output.txt"),
          data.table::fread) %>%
    data.table::rbindlist(use.names = TRUE) %>%
    data.table::fwrite("combined_output.txt", sep = "\t")

```
  4. Combine tables of binders and non-binders for Figures 2 and 3 from within R.
```
library(data.table)
library(magrittr)
library(uuid)

# combine our MHC binders tables
lapply(c("VASR_MHC_binders.txt",
            "Riaz_MHC_binders.txt",
            "Hellmann_MHC_binders.txt"),
            data.table::fread) %>%
      data.table::rbindlist(use.names = TRUE) %>%

      # add a UUID for each nmer to match BLAST results easily later on.
      .[, uuid := uuid::UUIDgenerate, by = 1:nrow(.)] %>%

      data.table::fwrite("../Figure2/all_MHC_binders_w_id.txt", sep = "\t")

# combine our Non-binders tables
lapply(c("VASR_non_binders.txt",
        "Riaz_non_binders.txt",
        "Hellmann_non_binders.txt"),
        data.table::fread) %>%

          data.table::rbindlist(use.names = TRUE) %>%

          data.table::fwrite("../Figure2/all_non_binders_w_names.txt", sep = "\t")
```
  5. From this point forward the analysis can be completed in any order. Start here to repeat the statistical analysis and figure generation from raw data without repeating the main neoantigen prediction analysis from the five datasets' input.

  6. Set the working directory to Figure 1 and run through the Fig1BC.R and Fig1D.R script. If you don't want to repeat the MHC affinity portion of the analysis from scratch, decompress the ensemble predictions data from the shell as below and start from line 76. You will first need to install [pixz](https://github.com/vasi/pixz), then:
```
pixz -d ensemble_pred_mtable.tpxz
tar -xvf ensemble_pred_mtable.tar
```
  7. Set the working directory to Figure 2. Run scripts Fig2A.R Fig2D.R, and Fig2CE_S2A.R. This will execute the analysis and generate figures 2A,C-E and S2A.

  8. Set the working directory to Figure 3. Run scripts Fig3A_S2B.R, Fig3CDE.R, and Fig3F.R. This will execute the analysis and generate figures 3A, C-F  and S2B.

  9. Set the working directory to Figure 4. Run scripts Fig4A.R and Fig4BC.R. This will execute the analysis and generate figures 4A-C. If you wish to repeat the analysis of all neopeptides independent of affinity, run script All_dissimilarity_9mers.R. Output from this script is also provided so this step is optional (Dissimilarity_9mers_patient_level.txt).

  10. Set the working directory to Figure S. Various scripts here generate analyses that were not part of the manuscript figure and may be run in any order. FigS1.R recreates the published benchmark analysis in Figure S1. FigS3D.R recreates Figure S3D. Alt_cutoff_contingency_tables.R generates Figures S2C and additional contingency tables at cutoffs of interest. Other scripts: *Keskin_analysis.R*, *Fritsch_Ott.R*, *Le_analysis.R*, *Chheda_H3.3K27M.R*, and *Balachandran_muc16.R* contain analysis of literature examples of neoantigens.

# Citation
Don't forget to cite *antigen.garnish* if you use it in your studies:

> Richman LP, Vonderheide RH, and Rech AJ. "Neoantigen dissimilarity to the self-proteome predicts immunogenicity and response to immune checkpoint blockade." *Cell Systems* 2019 *in press*

Again, please feel free to contact us with any questions about this analysis or *antigen.garnish* at [@leeprichman](https://github.com/leeprichman) or [leepr@upenn.edu](mailto:leepr@upenn.edu).
